File: Leetcode/104-Maximum-Depth-of-Binary-Tree/104-Maximum-Depth-of-Binary-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int maxDepthRec(TreeNode *root, int counter) {
        if (root == nullptr) {
            return counter;
        }

        counter++;
        return max(maxDepthRec(root->left, counter), maxDepthRec(root->right, counter));
    }
public:
    int maxDepth(TreeNode* root) {
        return maxDepthRec(root, 0);
    }
};

File: Leetcode/104-Maximum-Depth-of-Binary-Tree/104-Maximum-Depth-of-Binary-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        return recMaxDepth(root, 0);
    }

    private int recMaxDepth(TreeNode root, int currDepth) {
        if (root == null) {
            return currDepth;
        }

        currDepth++;

        return Math.max(recMaxDepth(root.left, currDepth), recMaxDepth(root.right, currDepth));
    }
}